tensorflow/core/kernels/stack.cc

```diff
@@ -72,15 +78,85 @@ class Stack : public ResourceBase {
       return errors::InvalidArgument("Stack[", stack_name_,
                                      "] is empty when calling Pop().");
     }
+    int index = stack_.size() - 1;
+    while (is_swapping_ins_[index]) {
+      cond_swapping_ins_[index].wait(l);
+      index = stack_.size() - 1;
+    }
     *value = stack_.back();
     stack_.pop_back();
     return Status::OK();
   }
 
+  Status SwapOut(TensorAndAllocation* value, Tensor* cpu_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *cpu_tensor;
+    value->swapped_to_cpu = true;
+
+    return Status::OK();
+  }
+
+  Status SwapIn(TensorAndAllocation* value, Tensor* device_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *device_tensor;
+    value->swapped_to_cpu = false;
+
+    return Status::OK();
+  }
+
+  bool GetTensorToSwapOut(
+      TensorAndAllocation** value,
+      std::function<bool(const TensorAndAllocation&)> cond) {
```

Comment:
The `cond` here actually is just the anonymous function:

```
[stack](const Stack::TensorAndAllocation& value) -> bool {
            return value.tensor.TotalBytes() > kCopyThreshold &&
                   stack->IsUsefulToSwap(value.tensor);
          }
```

found below. This is an adaptation to the check found in the pre-modified code.

Sorry if it looks complicated, I'm open to improvements.